lib/kafo/answer_file.rb

```diff
@@ -0,0 +1,68 @@
+require 'yaml'
+
+module Kafo
+  class AnswerFile
+
+    attr_reader :answers, :filename, :version
+
+    def initialize(answer_filename, version: 1, exit_handler: KafoConfigure, logger: KafoConfigure.logger)
+      @filename = answer_filename
+      @version = version.nil? ? 1 : version
+      @exit_handler = exit_handler
+      @logger = logger
+
+      begin
+        @answers = YAML.load_file(@filename)
+      rescue Errno::ENOENT
+        @exit_handler.exit(:no_answer_file) do
+          @logger.error "No answer file found at #{@filename}"
+        end
+      end
+
+      validate
+    end
+
+    def filename
+      @filename
+    end
+
+    def puppet_classes
+      @answers.keys.sort
+    end
+
+    def parameters_for_class(puppet_class)
+      if @version == 1
```

Comment:
Yea, thats a fair idea. I was not a big fan of the if-else, but it was the crude way my brain got around the problem. I am bouncing back and forth between this and a yet unopened commit to try to balance the setup for V2 (https://github.com/ehelms/kafo/commit/2b96d65a185fc585ebdaa508e76a648bc33105e1).

Reply:
Before a v2 we also need to introduce a migration DSL because now that also operates directly on the answers. Some things I thought about:

```ruby
# Ensure a module is present in the answers
def add_module(name, default=false)
end

# Ensure a module is absent in the answers
def remove_module(name)
end

# Ensure a module is enabled in the answers
def enable_module(name)
end

# Ensure a module is disabled in the answers
def disable_module(name)
end

# Set an answer for a module's parameter. If a module is disabled, it will be enabled
def set_answer(mod, parameter, value)
end

# Get an answer for a module's parameter. If a module is disabled or no answer for the parameter exists, it will return nil
def get_answer(mod, parameter)
end

# Clear an answer for a module's parameter. If a module is disabled, it does nothing
def unset_answer(mod, parameter)
end

# Migrate a module. The given block is only called if the module is enabled
# @example
#   migrate_module('puppet') do |mod|
#     current_autosign = mod['autosign']
#     unless current_autosign.is_a?(TrueClass) || current_autosign.is_a?(FalseClass)
#       mod.unset_answer('autosign')
#     end
#   end
#
# @example
#   migrate_module('foreman_proxy') do |mod|
#     mod.rename_parameter('puppetrun', 'puppet')
#     mod.rename_parameter('puppetrun_listen_on', 'puppet_listen_on')
#   end
def migrate_module(mod, &block)
end
```

I'm not entirely sure about the migrate_module name. I think the essence is that it operates on a module only if it's enabled and you get a module back which has some ways of expressing operations.

In https://github.com/theforeman/foreman-installer/pull/700 I tried to see if I could express the migrations and it looks like it. It feels more natural to me. I skipped a few complexer ones and all katello/foreman-proxy-content ones so I may have missed some things.

Reply:
I'll look this over. I wanted to make a note of something I wanted to discuss more as its something I find a bit odd and affects things. We draw some parallels to databases and migrations with the answer file yet we treat it like its both the end product and the baseline that migrations are layered on top of. I had been wondering if we ought to have this more formalized that really the answer file is the initial migration and migrations layer on top of that. I think this would mean we need a helper to calculate the answer file final state, but in a way we need that today as you can't see the affect of migrations either. This would also allow "squashing" migrations in the future similar to the Rails model.

Reply:
I think i'll need this change in place so that I have the interface to build the DSL into.